### PR TITLE
Make highlighting of deleted executables configurable

### PR DESCRIPTION
**Pull request by BenBE**
Addresses #383. Also updates documentation of the highlighting.

Should go in AFTER #388 due to code dependencies.

### CODE REVIEW

**Comment by fasterit** [comment]
Milestoned to 3.0.6 due to dependency on #388.
DLange still thinks this one need not be configurable. Users should know when executables have been deleted. Not wanting to know is ... a very odd us case for running a system monitoring application like htop.

**Comment by BenBE** [comment]
See also the related discussion in https://github.com/htop-dev/htop/issues/376#issuecomment-751249861 …

**Comment by milosivanovic** [comment]
> See also the related discussion in [#376 (comment)](https://github.com/htop-dev/htop/issues/376#issuecomment-751249861) …

I wasn't sure if the discussion should be here or in #376. I responded to @fasterit's comment there.

**Comment by BenBE** [comment]
> > See also the related discussion in [#376 (comment)](https://github.com/htop-dev/htop/issues/376#issuecomment-751249861) …
> 
> I wasn't sure if the discussion should be here or in #376. I responded to @fasterit's comment there.

Let's just keep the discussion central in #376. The "needs-discussion" label on this issue is mostly a team-internal marker for planning and coordinating things.

File: linux/LinuxProcess.c [hunk 1]
```diff
@@ -381,180 +402,139 @@ void LinuxProcess_makeCommandStr(Process* this) {
...
    const LinuxProcess *lp = (const LinuxProcess *)this;
    const LinuxProcessMergedCommand *mc = &lp->mergedCommand;
 
    int strStart = RichString_size(str);
 
-   int baseStart = strStart + lp->mergedCommand.baseStart;
-   int baseEnd = strStart + lp->mergedCommand.baseEnd;
-   int commStart = strStart + lp->mergedCommand.commStart;
-   int commEnd = strStart + lp->mergedCommand.commEnd;
+//   int commAttr = CRT_colors[Process_isUserlandThread(this) ? PROCESS_THREAD_COMM : PROCESS_COMM];
```
Comment (cgzones) [inline linux/LinuxProcess.c:509]: why keep this as comment?

File: linux/LinuxProcess.c [hunk 2]
```diff
@@ -381,180 +402,139 @@ void LinuxProcess_makeCommandStr(Process* this) {
...
 
-   int commAttr = CRT_colors[Process_isUserlandThread(this) ? PROCESS_THREAD_COMM : PROCESS_COMM];
+   const bool highlightBaseName = this->settings->highlightBaseName;
+   const bool highlightSeparator = true;
```
Comment (cgzones) [inline linux/LinuxProcess.c:512]: why a variable that is always `true`?

File: linux/LinuxProcess.c [hunk 3]
```diff
@@ -381,180 +402,139 @@ void LinuxProcess_makeCommandStr(Process* this) {
...
-            RichString_setAttrn(str, A_BOLD | commAttr, baseEnd, commEnd - baseEnd);
-         } else if (commEnd < baseEnd) {
-            RichString_setAttrn(str, A_BOLD | commAttr, commStart, commEnd - commStart);
-            RichString_setAttrn(str, A_BOLD | baseAttr, commEnd, baseEnd - commEnd);
-         } else {
-            // Actually should be highlighted commAttr, but marked baseAttr to reduce visual noise
-            RichString_setAttrn(str, A_BOLD | baseAttr, commStart, commEnd - commStart);
-         }
+      if (hl->flags & CMDLINE_HIGHLIGHT_FLAG_SEPARATOR)
+         if (!highlightSeparator)
```
Comment (cgzones) [inline linux/LinuxProcess.c:525]: maybe combine into `if ((hl->flags & CMDLINE_HIGHLIGHT_FLAG_SEPARATOR) && !highlightSeparator)`

File: Process.h [hunk 4]
```diff
@@ -241,6 +253,11 @@ static inline bool Process_isChildOf(const Process* this, pid_t pid) {
    return pid == Process_getParentPid(this);
 }
 
+#define CMDLINE_HIGHLIGHT_FLAG_SEPARATOR  0x00000001
```
[the remark this answers is not in the document]
Comment (BenBE) [reply Process.h:256]: Yes, these constants are used for the bitset with the coloring isntructions. May refactor to some enum later on.

File: Settings.h [hunk 5]
```diff
@@ -54,6 +54,7 @@ typedef struct Settings_ {
    bool hideKernelThreads;
    bool hideUserlandThreads;
    bool highlightBaseName;
+   bool highlightDeletedExe;
```
[the remark this answers is not in the document]
Comment (BenBE) [reply Settings.h:57]: --verbose?

File: Process.h [hunk 6]
```diff
@@ -96,14 +93,29 @@ typedef struct Process_ {
...
     * Process name including arguments.
     * Use Process_getCommand() for Command actually displayed.
     */
-   char* comm;
+   char* cmdline;
 
-   /* Offset in comm of the process basename */
-   int basenameOffset;
+   /* Offset in cmdline of the process basename */
+   int cmdlineBasenameOffset;
```
[the remark this answers is not in the document]
Comment (BenBE) [reply Process.h:106]: Still investigating that case. Some platforms use 0 while some places use -1 for non-existing offsets. Using `-1` overall makes some code places less manageable.

File: linux/LinuxProcess.h [hunk 7]
```diff
@@ -31,39 +31,39 @@ in the source distribution for its full text.
...
+   LinuxProcessCmdlineHighlight highlights[8]; /* which portions of cmdline to highlight */
+   bool separateComm : 1;   /* whether comm is a separate field */
+   bool unmatchedExe : 1;   /* whether exe matched with cmdline */
+   bool cmdlineChanged : 1; /* whether cmdline changed */
+   bool exeChanged : 1;     /* whether exe changed */
+   bool commChanged : 1;    /* whether comm changed */
+   bool prevMergeSet : 1;   /* whether showMergedCommand was set */
+   bool prevPathSet : 1;    /* whether showProgramPath was set */
+   bool prevCommSet : 1;    /* whether findCommInCmdline was set */
+   bool prevCmdlineSet : 1; /* whether findCommInCmdline was set */
```
[the remark this answers is not in the document]
Comment (BenBE) [reply linux/LinuxProcess.h:61]: That's inherited from the #388. Will need to fix there. This PR mostly is only the last 2 commits.

File: linux/LinuxProcess.c (hunk 1 above)
Comment (BenBE) [reply linux/LinuxProcess.c:509]: Probably for refactoring reference.

File: linux/LinuxProcess.c (hunk 2 above)
Comment (BenBE) [reply linux/LinuxProcess.c:512]: To have a reference for later, which places are affected when highlighting a separator. Will iron out once the overall code structure has been better streamlined overall.

File: linux/LinuxProcess.c (hunk 3 above)
Comment (BenBE) [reply linux/LinuxProcess.c:525]: intentional to illustrate the overall flag/condition structure.

Comment (cgzones) [reply linux/LinuxProcess.c]: fair enough

File: linux/LinuxProcess.c (hunk 2 above)
Comment (cgzones) [reply linux/LinuxProcess.c]: fair enough

**Review by cgzones — APPROVED** [review]

**Pull request merged**